### PR TITLE
Recommend Unique Action (RUA)

### PR DESCRIPTION
**Pull request by tcharding**
We want to recommend to the user what is the next action to take for a swap during the COMIT protocol.

Refund (and only refund) is returned once it is possible to refund, meaning that the HTLC has expired. This is done by checking the blockchain.

This PR is the result of the RUA shape up (@tcharding and @D4nte).

- Patch 1 and 2 are bug fixes to the actions logic
- Patch 3 fixes the e2e tests so that expiries are correct
- Patch 4 adds the RUA logic
- Patch 5 refactors the whole thing

Patch 5 is left separate to assist review, if this is not the case please say and we can squash 4 and 5 together.

This is v2 (total re-write) of https://github.com/comit-network/comit-rs/pull/2736.  Sorry for breaking the thread of context, seems we _still_ don't know how to use GitHub :)

This work is not tested any more than the current testing provided by the e2e tests. In particular, regression tests for this work are not added. There is a [tracking issue](https://github.com/comit-network/comit-rs/issues/2753) for adding testing (or fixing the API), we would like to merge this PR separately before starting work on that ticket.

### CODE REVIEW

File: cnd/src/http_api/routes.rs [hunk 1]
```diff
@@ -143,54 +181,99 @@ where
...
+                let time = match swap.beta_blockchain() {
+                    Blockchain::Bitcoin => facade.bitcoin_median_time_past().await?,
+                    Blockchain::Ethereum => facade.ethereum_latest_time().await?,
+                };
+                (expiry, time)
+            }
+        };
+
+        if expiry < blockchain_time {
+            tracing::debug!("We have decided it's time to refund.");
```
Comment (D4nte) [inline cnd/src/http_api/routes.rs:267]: You can remove this debug comment now or phrase it better (include the times for example)

Comment (tcharding) [reply cnd/src/http_api/routes.rs:267]: Cool, will remove.

File: cnd/src/http_api/herc20_halight/bob.rs [hunk 2]
```diff
@@ -218,3 +219,41 @@ impl From<BobSwap<asset::Erc20, asset::Bitcoin, herc20::Finalized, halight::Fina
...
+            } => Some(*expiry)
+        }
+    }
+}
+
+impl BetaAbsoluteExpiry
+    for BobSwap<asset::Erc20, asset::Bitcoin, herc20::Finalized, halight::Finalized>
+{
+    fn beta_absolute_expiry(&self) -> Option<Timestamp> {
+        None // No absolute expiry time for halight.
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/http_api/herc20_halight/bob.rs:257]: I like it like this.

File: api_tests/src/actors/swap_factory.ts [hunk 3]
```diff
@@ -273,8 +273,8 @@ function defaultHerc20EthereumErc20RequestParams(
 }
 
 function defaultHalightHanHerc20Expiries() {
-    const alphaAbsoluteExpiry = Math.round(Date.now() / 1000) + 120;
-    const betaAbsoluteExpiry = Math.round(Date.now() / 1000) + 60;
+    const alphaAbsoluteExpiry = Math.round(Date.now() / 1000) + 240;
```
[the remark this answers is not in the document]
Comment (tcharding) [reply api_tests/src/actors/swap_factory.ts:276]: Sure thing, thanks.

[the remark this answers is not in the document]
Comment (tcharding) [reply api_tests/src/actors/swap_factory.ts:276]: oh,  mergify is running already.

File: cnd/src/http_api/routes.rs (hunk 1 above)
Comment (tcharding) [reply cnd/src/http_api/routes.rs:267]: bors is running, leaving this for follow up if you feel strongly about it.

File: cnd/src/http_api/routes.rs [hunk 4]
```diff
@@ -154,137 +181,99 @@ where
...
+        + RefundAction
+        + Clone
+        + AlphaBlockchain
+        + BetaBlockchain
+        + AlphaAbsoluteExpiry
+        + BetaAbsoluteExpiry,
+{
+    if swap.init_action().is_ok() {
+        return Ok(Some(ActionName::Init));
     }
```
Comment (thomaseizinger) [inline cnd/src/http_api/routes.rs]: I was about to comment in the other patch that this logic is pretty confusing and thankfully, you are refactoring it here :D

Would have preferred to have this in one patch if patch-by-patch review is being recommended :)
As you once said @tcharding, "I don't care about the steps it took you to come to the solution" :)

File: cnd/src/http_api/protocol.rs [hunk 5]
```diff
@@ -36,6 +36,33 @@ pub trait BetaEvents {
...
+}
+
+/// Get the absolute expiry time for the beta protocol.
+pub trait BetaAbsoluteExpiry {
+    fn beta_absolute_expiry(&self) -> Option<Timestamp>;
+}
+
+/// Blockchains we currently support swaps on top of.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Blockchain {
```
[the remark this answers is not in the document]
Comment (tcharding) [reply cnd/src/http_api/protocol.rs:61]: My bad, I wondered when adding this why this identifier was non-existent :) Will do a follow up PR.

File: cnd/src/http_api/routes.rs (hunk 4 above)
Comment (tcharding) [reply cnd/src/http_api/routes.rs]: Yeah, I was lazy. The initial refactor of this function _should_ have been done as the initial patch and then only the additional logic would be in the final patch. But by the time I did the refactor I was too lazy to rebase and do it properly.

**Pull request merged**